### PR TITLE
US_Prams

### PR DESCRIPTION
**Pull request by HarishC727**
new file:   scripts/cdc_prams/input_files.py
	new file:   scripts/cdc_prams/process.py
	new file:   scripts/cdc_prams/statvar.py

### CODE REVIEW

File: scripts/cdc_prams/process.py [hunk 1]
```diff
@@ -0,0 +1,758 @@
...
+                       _TIME, _INSURANCE, _PROP4, _PV_PROP, _YEAR)
+# pylint: enable=import-error
+# pylint: enable=wrong-import-position
+_FLAGS = flags.FLAGS
+default_input_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                                  "input_files")
+flags.DEFINE_string("input_path", default_input_path, "Import Data File's List")
+
+
+def _multiline(df, flag):
```
Comment (AravindhVenkatesh) [inline scripts/cdc_prams/process.py:48]: Please provide different name for function.
Replace flag with geo in all instances. Make geo as enum with values National and State.
Incase flag is not used in function, then remove it.

File: scripts/cdc_prams/process.py [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+
+            if flag == "National":
+                # temp_df = temp_df.drop(columns=[
+                #     '2016_sampleSize', '2017_sampleSize', '2018_sampleSize',
+                #     '2019_sampleSize', '2020_sampleSize'
+                # ])
+                for year in range(2016, 2021):
+                    drop_columns.append(year + '_sampleSize')
+                temp_df = temp_df.drop(columns=drop_columns)
+            elif flag == "States":
```
Comment (AravindhVenkatesh) [inline scripts/cdc_prams/process.py:332]: use State

File: scripts/cdc_prams/process.py [hunk 3]
```diff
@@ -0,0 +1,758 @@
...
+            'newStatVar', '2016_CI', '2017_CI', '2018_CI', '2019_CI', '2020_CI',
+            'Overall_2020_CI'
+        ])
+        # Redifining columns
+    if flag == "National":
+        df = df[[
+            'Geo', 'SV', '2016_sampleSize', '2016_CI', '2017_sampleSize',
+            '2017_CI', '2018_sampleSize', '2018_CI', '2019_sampleSize',
+            '2019_CI', '2020_sampleSize', '2020_CI', 'ScalingFactor'
+        ]]
```
Comment (AravindhVenkatesh) [inline scripts/cdc_prams/process.py:261]: generate dynamically instead of hardcoding

File: scripts/cdc_prams/process.py [hunk 4]
```diff
@@ -0,0 +1,758 @@
...
+        # pylint: disable=R0912
+        # pylint: disable=R0915
+        mcf_nodes = []
+        dcid_nodes = {}
+        for sv in sv_names:
+            pvs = []
+            dcid = sv
+            sv_prop = [prop.strip() for prop in sv.split(" ")]
+            # Using statistical variable template
+            sv_pvs = deepcopy(DEFAULT_SV_PROP)
```
Comment (AravindhVenkatesh) [inline scripts/cdc_prams/process.py:517]: Add a comment why deepcopy is required

File: scripts/cdc_prams/process.py [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+            '2019_CI_PERCENT', '2019_CI_LOWER', '2019_CI_UPPER',
+            '2020_CI_PERCENT', '2020_CI_LOWER', '2020_CI_UPPER', 'ScalingFactor'
+        ]]
+        # The Distric of Columbia has characters : (.-.)
+        df['2018_CI_UPPER'] = df['2018_CI_UPPER'].replace('.', '0.0')
+        df['2018_CI_LOWER'] = df['2018_CI_LOWER'].replace('.', '0.0')
+    return df
+
+
+def _stat_var(df, flag):
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/process.py:278]: Renamed the function.

File: scripts/cdc_prams/process.py (hunk 2 above)
Comment (HarishC727) [reply scripts/cdc_prams/process.py:332]: Changes from States to State.

File: scripts/cdc_prams/process.py (hunk 3 above)
Comment (HarishC727) [reply scripts/cdc_prams/process.py:261]: Made it dynamic.

File: scripts/cdc_prams/process.py (hunk 1 above)
Comment (HarishC727) [reply scripts/cdc_prams/process.py:48]: Replaced flag with geo.

**Review by beets — COMMENTED** [review]
hi - sorry for the delay and i haven't been able to finish reviewing all the scripts. sending some preliminary feedback to unblock this.

File: scripts/cdc_prams/constants.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        "",
+    "CigaretteSmoking":
+        "",
+    "HookahInLast2Years":
+        "Last2YearsBeforePregnancy",
+    "HeavyDrinking":
+        "",
+    "CDC_SelfReportedDepression":
+        "",
+    "healthInsuranceStatus1MonthBeforePregnancyprivateinsurance":
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/constants.py:129]: Changed as per suggestion.

File: scripts/cdc_prams/constants.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        "HookahUsage",
+    "Obese":
+        "Obesity",
+    "healthInsuranceStatus1MonthBeforePregnancyprivateinsurance":
+        "WithPrivateHealthInsurance",
+    "healthInsuranceStatus1MonthBeforePregnancyMedicaid":
+        "WithMedicaid",
+    "healthInsuranceStatus1MonthBeforePregnancyNoInsurance":
+        "NoHealthInsurance",
+    "healthInsuranceStatusForPrenatalCareprivateinsurance":
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/constants.py:70]: Changed as per suggestion.

**Comment by beets** [comment]
Please also use [gitlfs](https://git-lfs.github.com/) for binary files and other large static files for testing. Thanks

File: scripts/cdc_prams/README.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        9. Oral Health.
+        10. Health Insurance Status One Month Before Pregnancy.
+        11. Health Insurance Status for Prenatal Care.
+        12. Health Insurance Status Postpartum.
+        13. Infant Sleep Practices.
+        14. Breastfeeding Practices.
+        
+
+### Download URL
+The data in .pdf formats are downloadable from https://www.cdc.gov/prams/prams-data/mch-indicators/states/pdf/2020/ -> 	State Name.
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/README.md:25]: Added an example url.

File: scripts/cdc_prams/download_input_files.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+import os
+from absl import app, flags
+# For import common.download
+# pylint: disable=import-error
+# pylint: disable=wrong-import-position
+# _COMMON_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+# sys.path.insert(1, _COMMON_PATH)
+from download import download_file
+# pylint: enable=import-error
+# pylint: enable=wrong-import-position
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/download_input_files.py:28]: Removed import errors.

File: scripts/cdc_prams/constants.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+                  "observationDate: C:US_Prams->Year\n"
+                  "scalingFactor: C:US_Prams->ScalingFactor\n"
+                  "value: C:US_Prams->Observation\n")
+
+DEFAULT_SV_PROP = {
+    "typeOf": "dcs:StatisticalVariable",
+    "populationType": "dcs:LiveBirthPregnancyEvent",
+}
+
+_PV_PROP = {
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/constants.py:39]: Added a comment.

File: scripts/cdc_prams/process.py [hunk 11]
```diff
@@ -0,0 +1,708 @@
...
+cleans them and generates the cleaned CSV, MCF and TMCF file.
+"""
+import os
+import sys
+import re
+from copy import deepcopy
+import pandas as pd
+import numpy as np
+from absl import app, flags
+# from util.alpha2_to_dcid import COUNTRY_MAP
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/process.py:25]: Deleted the line.

File: scripts/cdc_prams/process.py [hunk 12]
```diff
@@ -0,0 +1,708 @@
...
+import tabula as tb
+
+_CODEDIR = os.path.dirname(os.path.abspath(__file__))
+sys.path.insert(1, _CODEDIR)
+# pylint: disable=import-error
+# pylint: disable=wrong-import-position
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/process.py:31]: Removed the disable commands.

File: scripts/cdc_prams/constants.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+        "",
+    "ECigaretteSmoking":
+        "Smoking",
+    "CigaretteSmoking":
+        "Smoking",
+    "SampleSize_Count":
+        "",
+    "Percent":
+        "",
+    "ConfidenceIntervalLowerLimit_Count":
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/constants.py:92]: Added a comment.

File: scripts/cdc_prams/download.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    for download_file_url in input_url:
+        # Example, file_name_with_compression_ext: NHTS_2009_transfer_AL.zip
+        file_name_with_compression_ext = os.path.basename(download_file_url)
+        # Example, file_name_without_compression_ext: NHTS_2009_transfer_AL.txt
+        file_name_without_compression_ext = os.path.splitext(
+            file_name_with_compression_ext)[0]
+        out_file = path + os.sep + file_name_with_compression_ext
+        print(download_file_url)
+        req = requests.get(download_file_url)
+        if ".zip" not in download_file_url:
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/download.py:49]: Used endswith.

File: scripts/cdc_prams/README.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+#### MCFs and Template MCFs
+- [output/PRAMS.mcf]
+- [output/PRAMS.tmcf]
+
+### Running Tests
+
+Run the test cases
+
+`python3 -m unittest scripts/cdc_prams/process_test.py`
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/README.md:70]: Can we modify the requirement.txt file? To run the import we need tabula-py package.

File: scripts/cdc_prams/process.py [hunk 16]
```diff
@@ -0,0 +1,708 @@
...
+        'Experienced IPV during pregnancy by a husband or partner'
+    ]
+    # Converting the multiline statistical variable into single line
+    # statistical variable
+    for line in multiline:
+        for i in range(len(df)):
+            if df.loc[i, 'statVar'] == line:
+                df.loc[i,'statVar'] = \
+                    f"{df.loc[i,'statVar']}{' '}{df.loc[i + 2,'statVar']}"
+                df.loc[i, '2016_sampleSize'] = df.loc[i + 1, 'statVar']
```
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/process.py:63]: Made year as flag.

File: scripts/cdc_prams/constants.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (beets) [reply scripts/cdc_prams/constants.py:40]: What I meant was to comment what the dictionaries are for.. e.g. "Holds mappings of __ to __. Used for __."

File: scripts/cdc_prams/constants.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (beets) [reply scripts/cdc_prams/constants.py:94]: I don't see the comment that explains this..

File: scripts/cdc_prams/download_input_files.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (beets) [reply scripts/cdc_prams/download_input_files.py:28]: thanks for resolving. can you remove the pylint disables?

**Review by beets — COMMENTED** [review]
thanks for the updates! some more comments below. please also use gitlfs for the pdfs and large test files

File: scripts/cdc_prams/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/README.md:71]: Updated the requirement.txt

File: scripts/cdc_prams/download_input_files.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/download_input_files.py:28]: Removed it.

File: scripts/cdc_prams/constants.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/constants.py:94]: Updated the comment.

File: scripts/cdc_prams/constants.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HarishC727) [reply scripts/cdc_prams/constants.py:40]: Updated the comment.

**Review by beets — APPROVED** [review]
thanks for the updates.

**Pull request merged**